python/cudf/cudf/core/indexed_frame.py

```diff
@@ -175,6 +188,152 @@ def to_dict(self, *args, **kwargs):  # noqa: D102
             "`.to_pandas().to_dict()` to construct a Python dictionary."
         )
 
+    @property
+    def _num_rows(self) -> int:
+        # Important to use the index because the data may be empty.
+        return len(self._index)
+
+    @classmethod
+    def _from_data(
+        cls,
+        data: MutableMapping,
+        index: Optional[BaseIndex] = None,
+        *args,
+        **kwargs,
+    ):
+        out = super()._from_data(data, *args, **kwargs)
+        out._index = RangeIndex(out._data.nrows) if index is None else index
+        return out
+
+    @classmethod
+    @_cudf_nvtx_annotate
+    def _from_columns(
+        cls,
+        columns: List[ColumnBase],
+        column_names: List[str],
+        index_names: Optional[List[str]] = None,
+    ):
+        """Construct a `Frame` object from a list of columns.
+
+        If `index_names` is set, the first `len(index_names)` columns are
+        used to construct the index of the frame.
+        """
+        data_columns = columns
+
+        n_index_columns = len(index_names) if index_names else 0
+        index_columns = columns[:n_index_columns]
+        data_columns = columns[n_index_columns:]
+
+        out = super()._from_columns(data_columns, column_names)
+
+        if index_names is not None:
+            out._index = cudf.core.index._index_from_columns(index_columns)
+            if isinstance(out._index, cudf.MultiIndex):
+                out._index.names = index_names
+            else:
+                assert len(index_names) == 1
```

Comment:
raise an explicit error with a message here? But the question also is, is this check necessary?

Reply:
I think no, we can remove this entirely. This is an internal function and developers can figure out what breaks if they need to.